FEATURE: Overhaul content stream pruner to work on event stream

mhsdesign opened the pull request:
Bugfix: Harden content stream pruner and remove `removed` flag from content streams

Resolves: #4913
Resolves: https://github.com/neos/neos-development-collection/issues/5220

With #5096 i introduced the `ContentStream::removed` flag and got rid of the sql query `findUnusedAndRemovedContentStreamIds` and calculated this information in php: https://github.com/neos/neos-development-collection/commit/041f23926f41fa0ad2dad6e62f16efb45aa2e836
While i [_thought_](https://github.com/neos/neos-development-collection/pull/5096#discussion_r1803786984) this was a good idea, it seems it was a right step in the right direction but we can improve on that:
- the removed flag is a bad idea, as it will complicate checking if a content stream now really exists
  - `$commandHandlingDependencies->contentStreamExists()` is allowed to return false if a contentstream existed once, as we fork with `ExpectedVersion::NO_STREAM()` and thus would still check that the fork is REALLY the first event. So any "hacked" case like deleting a content stream and using the same again to create a workspace is safe.

Additionally as part of the migration for the repaired publishing https://github.com/neos/neos-development-collection/pull/5301 we require the dangling content streams to be pruned. This is currently unstable as we rely on the projection to be up to date which is not the case for users of beta14 if they run into the publish bug.

For this reason we iterate over the events of the event store directly - filtering already the desired event types to limit the set - and build up the state of which content streams were removed, are no longer in use or in use by workspace.
That allows us to remove the same logic from the projection.

The change will also remove the left-over content stream commands that are not required by the workspace command handler anymore, by publishing events directly to replace their use in tests and in the pruner.


### New command `flow contentstream:status`

Detects if dangling content streams exists and which content streams could be pruned from the event stream

**Dangling content streams**

Content streams that are not removed via the event ContentStreamWasRemoved and are not in use by a workspace
(not a current's workspace content stream).

Previously before Neos 9 beta 15 (#5301), dangling content streams were not removed during publishing, discard or rebase.

```
./flow contentStream:removeDangling
```

**Pruneable content streams**

Content streams that were removed ContentStreamWasRemoved e.g. after publishing, and are not required for a full
replay to reconstruct the current projections state. The ability to reconstitute a previous state will be lost.

```
./flow contentStream:pruneRemovedFromEventStream
```

**_Example output_**

```
Dangling content streams that are not removed (ContentStreamWasRemoved) and not in use by workspace:
  id: c94b2a06-bbb0-4892-a92c-405baa0dc03a no longer in use
  id: 079ff29d-6900-4108-9eba-83c127da00e0 no longer in use
  id: 77769dc0-7b11-4405-ba6d-da4ad45b9dc2 no longer in use
  id: 7f64b63f-46c0-4d2d-8e2b-fdf73118aeb2 no longer in use
  id: e542b7d2-a7c1-4bd8-b02a-3e8450829965 no longer in use
  id: 7c41e343-bdff-4719-9dc7-8763cbb9d3db no longer in use
  id: 81a50ad4-41bb-43b9-8e2a-fac6b9beb3ee temporary forked at 2024-04-19 19:49
  id: 0db932cb-13bc-4f38-a4de-d0c304c7305f temporary forked at 2024-04-19 19:50
  id: ce276d40-3cba-4295-ab5c-c2d63dc8207f no longer in use
....

Removed content streams that can be pruned from the event store
  id: d1ce5fe8-9360-4886-9824-1788f2c0a327 previous state: forked
  id: 5771cc51-fb30-4571-8121-1c5c1d80dee8 previous state: no longer in use
  id: 2da82adc-3ec5-40fb-b4e3-977a896ae512 previous state: no longer in use
  id: 58c3a876-c8cf-460f-8ace-5cb9554c076a previous state: no longer in use
  id: 6541e994-9bf8-4df6-a071-bd51f36355f2 previous state: forked
  id: ca8357c7-2b05-4361-aa67-2102ec1969bf previous state: no longer in use
  id: 7a973dc1-f12e-4182-80ac-760a4fd7ec91 previous state: forked
  id: a749762d-b743-4368-8600-20d5d7658bd5 previous state: no longer in use
  id: 419102db-b28d-4cde-85dd-446ae2a163a8 previous state: forked
```

### Changes to `flow contentStream:removeDangling` (previously `contentStream:prune`)

The command was renamed to better distinct between pruning and removing.
We will now always include temporary content streams like `FORKED` or `CREATED` in the removal if they are older than 1 day. See `--remove-temporary-before=-1day` option. That way we can be sure not to remove temporary content streams of a users current action.

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

- [x] remove legacy rebased failed state from documentation
- [x] legacy rebased failed event is correctly handled 
- [x] content stream pruner works fully on event streams
- [x] content stream will be removed in projection instead of removed flag
- [x] content stream state/status was removed from the projection
- [x] content stream close status will now be represented in closed column
- [x] low level content stream commands were removed and so the content stream command handler


<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Remove `removed` state from ContentStream model and database

Followup to

https://github.com/neos/neos-development-collection/commit/041f23926f41fa0ad2dad6e62f16efb45aa2e836

By diffing `findAllContentStreamEventNames` with the `$transitiveUsedStreams` we can remove the `removed` flag from content streams, which would just introduce to much complexity.

mhsdesign committed:
BUGFIX: 4913 make `pruneAll` independent of projection

mhsdesign committed:
BUGFIX: Also remove closed content streams in prune

bwaidelich commented on Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php:
```diff
@@ -129,13 +131,29 @@ private function findUnusedAndRemovedContentStreams(): ContentStreams
+        return $removedContentStreams;
+    }
+
+    /**
+     * @return list<StreamName>
+     */
+    private function findAllContentStreamEventNames(): array
+    {
+        $events = $this->eventStore->load(VirtualStreamName::forCategory(ContentStreamEventStreamName::EVENT_STREAM_NAME_PREFIX));
+        $allContentStreamEventStreamNames = [];
+        foreach ($events as $eventEnvelope) {
+            $allContentStreamEventStreamNames[$eventEnvelope->streamName->value] = true;
+        }
+        return array_map(StreamName::fromString(...), array_keys($allContentStreamEventStreamNames));
```
more of a nit pick, but "ContentStreamEventNames" is not correct.
What about doing the mapping to the `ContentStreamId` here like so:

```suggestion
    /**
     * @return list<ContentStreamId>
     */
    private function findAllContentStreamIds(): array
    {
        $events = $this->eventStore->load(VirtualStreamName::forCategory(ContentStreamEventStreamName::EVENT_STREAM_NAME_PREFIX));
        $allContentStreamEventStreamNames = [];
        foreach ($events as $eventEnvelope) {
            $allContentStreamEventStreamNames[$eventEnvelope->streamName->value] = true;
        }
        return array_map(static fn (string $streamName) => substr($streamName, strlen(ContentStreamEventStreamName::EVENT_STREAM_NAME_PREFIX)), array_keys($allContentStreamEventStreamNames));
```
?

I know, this means that we do a couple of string operations for content streams that are not relevant above, but I don't think that this is measurable

mhsdesign replied:
the thing is we convert it later to `eventStream` things again ... so i rather keep the return value this way... maybe we can find a better name for `findAllContentStreamEventNames`?

bwaidelich reviewed (commented):
Thanks!
I don't have the brains anymore to do a proper review right now, so just a nitpicky-comment for now. But it looks great at first sight already!

mhsdesign committed:
TASK: Remove obsolete `REBASE_ERROR` content stream state

Via #4965 the status REBASE_ERROR is obsolete.
Instead of still using this status on replay we mimic the new logic:
> In case of a [rebase error}, reopen the old content stream and remove the newly created

Additionally, to not break `findContentStreams` when fetching all content streams we make sure that the streams with `REBASE_ERROR` are transformed to `NO_LONGER_IN_USE` so they can be cleaned up.

Related https://github.com/neos/neos-development-collection/issues/5101

mhsdesign commented on Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php:
```diff
@@ -73,28 +75,28 @@ public function prune(bool $removeTemporary = false): iterable
+    /**
+     * @return list<StreamName>
+     */
+    private function findUnusedAndRemovedContentStreamIds(): array
```
todo maybe not use the projection here at all to ensure the prune can always be run?

mhsdesign commented on Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php:
```diff
@@ -73,28 +75,28 @@ public function prune(bool $removeTemporary = false): iterable
-            $this->eventStore->deleteStream($streamName);
+        $removedContentStreams = $this->findUnusedAndRemovedContentStreamIds();
+        foreach ($removedContentStreams as $removedContentStreamName) {
+            $this->eventStore->deleteStream($removedContentStreamName);
```
todo do we need to delete workspace streams

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/harden-content-stream-pruner

mhsdesign committed:
TASK: Use `ContentStreamWasRemoved` event instead of `RemoveContentStream`

the command will be removed in the future

mhsdesign committed:
TASK: Build content stream for pruning state by reading the event stream

Otherwise, its not possible to prune if the projections are out of date or broken

mhsdesign committed:
TASK: Harden `contentstream:prune` to work on event stream and catch failed catchup

mhsdesign committed:
Linte is basically ente with i

mhsdesign committed:
TASK: Also handle `WorkspaceWasPublished` event

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/harden-content-stream-pruner

mhsdesign committed:
TASK: Prune also dangling streams of legacy `WorkspaceRebaseFailed` if existing

kitsunet reviewed (approved):
this works fine for me, I checked the status cases and they all make sense to me.

mhsdesign committed:
TASK: Remove now obsolete content stream status from projection

With the pruner now building up the state from scratch we dont need to project it any further.

The only exception is the content stream state that wasnt one to begin with: the closed state. It will get its own column.

mhsdesign committed:
TASK: Ensure that `cr:prune` prunes also workspaces cleanly

mhsdesign committed:
TASK: Avoid use of `CloseContentStream` command in tests and use event

Also as we dont expose the command anymore we can get rid of the constraint checks and:

> the command CloseContentStream is executed with payload and exceptions are caught

mhsdesign committed:
TASK: Remove legacy low level ContentStream commands and command handler :D

mhsdesign committed:
BUGFIX: ContentStream pruner mark `RootWorkspaceWasCreated` as in use

mhsdesign commented on Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php:
```diff
@@ -132,10 +173,159 @@ private function findUnusedAndRemovedContentStreams(): ContentStreams
+        foreach ($workspaceEvents as $eventEnvelope) {
+            $domainEvent = $this->eventNormalizer->denormalize($eventEnvelope->event);
+
+            switch ($domainEvent::class) {
```
upsi @kitsunet the `RootWorkspaceWasCreated` event was kind-of ignored here so the content stream was pruned in bernhards case is guess :O ... maybe we should backup the events before or at least ask for confirmation?

mhsdesign committed:
TASK: Always remove all dangling content streams

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into bugfix/harden-content-stream-pruner

mhsdesign committed:
BUGFIX: Mark workspace outdated if its source content stream was removed

During rebase of the base workspace, its content stream will be now fully removed from the projection (no soft removal) thats why we have to adjust the calculation to not rely on the select being `null` ONLY for when it has no base.
Its also null if the source content stream was removed.

```
001 Scenario: Publishing to the root workspace render dependents outdated # Features/Workspaces/WorkspaceState.feature:71
      Then workspace user-ws-two has status OUTDATED
```

mhsdesign committed:
TASK: `Workspace` to use static `create` like `Node` and other read models

mhsdesign commented on Neos.ContentRepository.Core/Classes/Feature/ContentStreamEventStreamName.php:
```diff
@@ -41,6 +41,15 @@ public static function isContentStreamStreamName(StreamName $streamName): bool
         return str_starts_with($streamName->value, self::EVENT_STREAM_NAME_PREFIX);
     }
 
+    public static function extractContentStreamId(StreamName $streamName): ContentStreamId
+    {
+        [$prefix, $contentStreamId] = explode(':', $streamName->value, 2);
+        if (($prefix . ':') !== self::EVENT_STREAM_NAME_PREFIX) {
+            throw new \InvalidArgumentException(sprintf('Stream name is not a content stream event stream "%s"', $streamName->value));
+        }
+        return ContentStreamId::fromString($contentStreamId);
+    }
```
unused can be reverted

mhsdesign commented on Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php:
```diff
@@ -24,47 +41,59 @@ class ContentStreamPruner implements ContentRepositoryServiceInterface
-     * @param bool $removeTemporary if TRUE, will delete ALL content streams not bound to a workspace
-     * @return iterable<int,ContentStreamId> the identifiers of the removed content streams
+     * @deprecated with Neos 9 beta 15, only used to migrate from earlier versions
      */
-    public function prune(bool $removeTemporary = false): iterable
+    public function removeDangelingContentStreams(\Closure $outputFn): void
```
todo, this could be part of some integration check rather as with the new publishing one will never run into this case

mhsdesign replied:
it cannot be part of the projection integrity things as ... well ... that is only for projections and if we forgot to cleanup after publishing its rather a core thing. Instead `flow contentStream:status` will help out ;)

mhsdesign commented on Neos.ContentRepository.Core/Classes/Service/ContentStreamPruner.php:
```diff
@@ -76,30 +105,37 @@ public function prune(bool $removeTemporary = false): iterable
+     * @return list<ContentStreamId> the removed content streams
      */
-    public function pruneRemovedFromEventStream(): ContentStreams
+    public function pruneRemovedFromEventStream(): array
```
todo maybe one wants full control over what will be removed and why? Should we introduce a dry mode? So one calls it once dry, then check and than reruns ... or we could backup the events table?

kitsunet replied:
I'd  say dry run seems fine? Could be a follow up though

mhsdesign replied:
build a little `contentstream:status` utility, and will add a warning to the prune commands that you might need to check with `contentstream:status` first and they should not be run if the system is in use as then a "forked" content stream might really appear for just a second ...

```
flow contentstream:status                     
Content streams that are not removed (ContentStreamWasRemoved) and not in use by workspace:
  id: 81a50ad4-41bb-43b9-8e2a-fac6b9beb3ee reason for removal: forked (temporary state)
  id: 0db932cb-13bc-4f38-a4de-d0c304c7305f reason for removal: forked (temporary state)
  id: 6ddb063c-ff46-4a12-8bc2-7bd701567b3b reason for removal: forked (temporary state)
  id: 3058560a-348e-497b-93a5-95418d49a54a reason for removal: forked (temporary state)
  id: 2a36581d-3c40-4289-99a7-deba45a052ef reason for removal: no longer in use
To remove the dangeling streams from the projections please run ./flow contentstream:prune
Then they are ready for removal from the event store

Removed content streams that can be pruned from the event store
  id: 310d09ef-4b5e-412c-9c08-9602ef05b85e previous state: no longer in use
  id: 5aa33bdd-95c2-4db3-98c6-5fd270a128b1 previous state: forked (temporary state)
To prune the removed streams from the event store run ./flow contentstream:pruneremovedfromeventstream
Then they are indefinitely pruned from the event store
```

mhsdesign commented on Neos.ContentRepositoryRegistry/Classes/Command/ContentStreamCommandController.php:
```diff
@@ -27,7 +27,7 @@ class ContentStreamCommandController extends CommandController
      *       `./flow contentStream:prune`.
      *
-     * By default, only content streams in STATE_NO_LONGER_IN_USE and STATE_REBASE_ERROR will be removed.
+     * By default, only content streams that are NO_LONGER_IN_USE will be removed.
```
todo update comment

mhsdesign committed:
TASK: Introduce `contentstream:status`

mhsdesign committed:
TASK: Only remove dangling streams `contentstream:removeDangling` that are older than x

That way we ensure this command can be run during production and doent kill any workspaces that are *just* created

mhsdesign committed:
TASK: Remove obsolete explicit pruning (I prune unused content streams) from tests

that is part of the publishing now

mhsdesign committed:
TASK: Adjust documentation and add force flag

mhsdesign committed:
TASK: Remove obsolete `extractContentStreamId`

mhsdesign committed:
TASK: Minor cleanup in docs and introduce `isDangling`

mhsdesign committed:
TASK: Test content stream pruner status output

mhsdesign commented:
I tested out that one can use 

```
./flow migrateevents:backup
./flow contentStream:removeDangling
./flow contentStream:pruneRemovedFromEventStream
./flow cr:projectionReplayAll
```

to recover from https://github.com/neos/neos-development-collection/issues/5327 as the content streams with the fake events (with workspace=live) will just be pruned.

Does anyone else wants to have a final look? (And try it out yourself?)
cc @bwaidelich cc @kitsunet :D

mhsdesign committed:
TASK: Introduce simple `flow migrateevents:backup`

mhsdesign committed:
TASK: Adjust variables and namings

mhsdesign committed:
TASK: Validate `--remove-temporary-before`

kitsunet reviewed (approved):
Great job, this should do well!
